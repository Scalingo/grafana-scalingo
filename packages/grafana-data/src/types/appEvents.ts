import { eventFactory } from './utils';

export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace AppEvents {
  export type AlertPayload = [string, string?];

  export const alertSuccess = eventFactory<AlertPayload>('alert-success');
  export const alertWarning = eventFactory<AlertPayload>('alert-warning');
  export const alertError = eventFactory<AlertPayload>('alert-error');
}
