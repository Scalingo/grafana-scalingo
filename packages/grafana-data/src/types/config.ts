import { SystemDateFormatSettings } from '../datetime';
import { MapLayerOptions } from '../geo/layer';
import { GrafanaTheme2 } from '../themes';

import { DataSourceInstanceSettings } from './datasource';
import { FeatureToggles } from './featureToggles.gen';
import { PanelPluginMeta } from './panel';
import { GrafanaTheme } from './theme';

import { NavLinkDTO, OrgRole } from '.';

/**
 * Describes the build information that will be available via the Grafana configuration.
 *
 * @public
 */
export interface BuildInfo {
  version: string;
  commit: string;
  env: string;
  edition: GrafanaEdition;
  latestVersion: string;
  hasUpdate: boolean;
  hideVersion: boolean;
}

/**
 * @internal
 */
export enum GrafanaEdition {
  OpenSource = 'Open Source',
  Pro = 'Pro',
  Enterprise = 'Enterprise',
}

/**
 * Describes the license information about the current running instance of Grafana.
 *
 * @public
 */
export interface LicenseInfo {
  expiry: number;
  licenseUrl: string;
  stateInfo: string;
  edition: GrafanaEdition;
  enabledFeatures: { [key: string]: boolean };
  trialExpiry?: number;
}

/**
 * Describes Sentry integration config
 *
 * @public
 */
export interface SentryConfig {
  enabled: boolean;
  dsn: string;
  customEndpoint: string;
  sampleRate: number;
}

/**
 * Describes the plugins that should be preloaded prior to start Grafana.
 *
 * @public
 */
export type PreloadPlugin = {
  path: string;
  version: string;
};

/** Supported OAuth services
 *
 * @public
 */
export type OAuth =
  | 'github'
  | 'gitlab'
  | 'google'
  | 'generic_oauth'
  // | 'grafananet' Deprecated. Key always changed to "grafana_com"
  | 'grafana_com'
  | 'azuread'
  | 'okta';

/** Map of enabled OAuth services and their respective names
 *
 * @public
 */
export type OAuthSettings = Partial<Record<OAuth, { name: string; icon?: string }>>;

/** Current user info included in bootData
 *
 * @internal
 */
export interface CurrentUserDTO {
  isSignedIn: boolean;
  id: number;
  externalUserId: string;
  login: string;
  email: string;
  name: string;
  lightTheme: boolean;
  orgCount: number;
  orgId: number;
  orgName: string;
  orgRole: OrgRole | '';
  isGrafanaAdmin: boolean;
  gravatarUrl: string;
  timezone: string;
  weekStart: string;
  locale: string;
  permissions?: Record<string, boolean>;
}

/** Contains essential user and config info
 *
 * @internal
 */
export interface BootData {
  user: CurrentUserDTO;
  settings: GrafanaConfig;
  navTree: NavLinkDTO[];
  themePaths: {
    light: string;
    dark: string;
  };
}

/**
 * Describes all the different Grafana configuration values available for an instance.
 *
 * @internal
 */
export interface GrafanaConfig {
  datasources: { [str: string]: DataSourceInstanceSettings };
  panels: { [key: string]: PanelPluginMeta };
  minRefreshInterval: string;
  appSubUrl: string;
  windowTitlePrefix: string;
  buildInfo: BuildInfo;
  newPanelTitle: string;
  bootData: BootData;
  externalUserMngLinkUrl: string;
  externalUserMngLinkName: string;
  externalUserMngInfo: string;
  allowOrgCreate: boolean;
  disableLoginForm: boolean;
  defaultDatasource: string;
  alertingEnabled: boolean;
  alertingErrorOrTimeout: string;
  alertingNoDataOrNullValues: string;
  alertingMinInterval: number;
  authProxyEnabled: boolean;
  exploreEnabled: boolean;
  helpEnabled: boolean;
  profileEnabled: boolean;
  ldapEnabled: boolean;
  sigV4AuthEnabled: boolean;
  samlEnabled: boolean;
  autoAssignOrg: boolean;
  verifyEmailEnabled: boolean;
  oauth: OAuthSettings;
  disableUserSignUp: boolean;
  loginHint: string;
  passwordHint: string;
  loginError?: string;
  navTree: any;
  viewersCanEdit: boolean;
  editorsCanAdmin: boolean;
  disableSanitizeHtml: boolean;
  liveEnabled: boolean;
  theme: GrafanaTheme;
  theme2: GrafanaTheme2;
  pluginsToPreload: PreloadPlugin[];
  featureToggles: FeatureToggles;
  licenseInfo: LicenseInfo;
  http2Enabled: boolean;
  dateFormats?: SystemDateFormatSettings;
  sentry: SentryConfig;
  customTheme?: any;
  geomapDefaultBaseLayer?: MapLayerOptions;
  geomapDisableCustomBaseLayer?: boolean;
  unifiedAlertingEnabled: boolean;
  angularSupportEnabled: boolean;
  feedbackLinksEnabled: boolean;
}
