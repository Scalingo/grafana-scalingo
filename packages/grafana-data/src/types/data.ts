export type KeyValue<T = any> = { [s: string]: T };

export enum LoadingState {
  NotStarted = 'NotStarted',
  Loading = 'Loading',
  Streaming = 'Streaming',
  Done = 'Done',
  Error = 'Error',
}

export interface QueryResultMeta {
  [key: string]: any;

  // Match the result to the query
  requestId?: string;

  // Used in Explore for highlighting
  searchWords?: string[];

  // Used in Explore to show limit applied to search result
  limit?: number;

  // HACK: save the datassource name in the meta so we can load it from the response
  // we should be able to find the datasource from the refId
  datasource?: string;

  // DatasSource Specific Values
  custom?: Record<string, any>;
}

export interface QueryResultBase {
  /**
   * Matches the query target refId
   */
  refId?: string;

  /**
   * Used by some backend data sources to communicate back info about the execution (generated sql, timing)
   */
  meta?: QueryResultMeta;
}

export interface Labels {
  [key: string]: string;
}

export interface Column {
  text: string; // For a Column, the 'text' is the field name
  filterable?: boolean;
  unit?: string;
}

export interface TableData extends QueryResultBase {
  name?: string;
  columns: Column[];
  rows: any[][];
  type?: string;
}

export type TimeSeriesValue = number | null;

export type TimeSeriesPoints = TimeSeriesValue[][];

export interface TimeSeries extends QueryResultBase {
  target: string;
  datapoints: TimeSeriesPoints;
  unit?: string;
  tags?: Labels;
}

export enum NullValueMode {
  Null = 'null',
  Ignore = 'connected',
  AsZero = 'null as zero',
}

export interface AnnotationEvent {
  id?: string;
  annotation?: any;
  dashboardId?: number;
  panelId?: number;
  userId?: number;
  login?: string;
  email?: string;
  avatarUrl?: string;
  time?: number;
  timeEnd?: number;
  isRegion?: boolean;
  title?: string;
  text?: string;
  type?: string;
  tags?: string[];

  // Currently used to merge annotations from alerts and dashboard
  source?: any; // source.type === 'dashboard'
}
