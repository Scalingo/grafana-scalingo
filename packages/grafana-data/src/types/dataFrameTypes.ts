/**
 * See also:
 * https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/frame_type.go
 *
 * @public
 */
export enum DataFrameType {
  TimeSeriesWide = 'timeseries-wide',
  TimeSeriesLong = 'timeseries-long',
  TimeSeriesMany = 'timeseries-many',

  /**
   * First field is X, the rest are bucket values
   */
  HeatmapBuckets = 'heatmap-buckets',

  /**
   * Explicit fields for:
   *  xMin, yMin, count, ...
   *
   * All values in the grid exist and have regular spacing
   */
  HeatmapScanlines = 'heatmap-scanlines',

  /** Directory listing */
  DirectoryListing = 'directory-listing',
}
