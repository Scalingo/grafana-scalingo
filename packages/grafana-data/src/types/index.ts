export * from './data';
export * from './dataFrame';
export * from './dataLink';
export * from './logs';
export * from './navModel';
export * from './select';
export * from './time';
export * from './thresholds';
export * from './utils';
export * from './valueMapping';
export * from './displayValue';
export * from './graph';
export * from './ScopedVars';
export * from './transformations';
export * from './fieldOverrides';
export * from './vector';
export * from './app';
export * from './datasource';
export * from './panel';
export * from './plugin';
export * from './thresholds';
export * from './fieldColor';
export * from './theme';
export * from './orgs';
export * from './flot';

export { AppEvent, AppEvents } from './appEvents';
export { PanelEvents } from './panelEvents';
