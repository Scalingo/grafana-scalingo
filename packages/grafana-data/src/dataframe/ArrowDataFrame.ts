import { DataFrame, FieldType, Field, Vector } from '../types';
import {
  Table,
  ArrowType,
  Builder,
  Vector as ArrowVector,
  Float64,
  DataType,
  Utf8,
  TimestampMillisecond,
  Bool,
  Column,
} from 'apache-arrow';

export interface ArrowDataFrame extends DataFrame {
  table: Table;
}

export function base64StringToArrowTable(text: string): Table {
  const b64 = atob(text);
  const arr = Uint8Array.from(b64, c => {
    return c.charCodeAt(0);
  });
  return Table.from(arr);
}

function valueOrUndefined(val?: string) {
  return val ? val : undefined;
}

function parseOptionalMeta(str?: string): any {
  if (str && str.length && str !== '{}') {
    try {
      return JSON.parse(str);
    } catch (err) {
      console.warn('Error reading JSON from arrow metadata: ', str);
    }
  }
  return undefined;
}

export function arrowTableToDataFrame(table: Table): ArrowDataFrame {
  const fields: Field[] = [];

  for (let i = 0; i < table.numCols; i++) {
    const col = table.getColumnAt(i);
    if (col) {
      const schema = table.schema.fields[i];
      let type = FieldType.other;
      const values: Vector<any> = col;
      switch ((schema.typeId as unknown) as ArrowType) {
        case ArrowType.Decimal:
        case ArrowType.Int:
        case ArrowType.FloatingPoint: {
          type = FieldType.number;
          break;
        }
        case ArrowType.Bool: {
          type = FieldType.boolean;
          break;
        }
        case ArrowType.Timestamp: {
          type = FieldType.time;
          break;
        }
        case ArrowType.Utf8: {
          type = FieldType.string;
          break;
        }
        default:
          console.log('UNKNOWN Type:', schema);
      }

      fields.push({
        name: col.name,
        type,
        values,
        config: parseOptionalMeta(col.metadata.get('config')) || {},
        labels: parseOptionalMeta(col.metadata.get('labels')),
      });
    }
  }
  const meta = table.schema.metadata;
  return {
    fields,
    length: table.length,
    refId: valueOrUndefined(meta.get('refId')),
    name: valueOrUndefined(meta.get('name')),
    meta: parseOptionalMeta(meta.get('meta')),
    table,
  };
}

function toArrowVector(field: Field): ArrowVector {
  // OR: Float64Vector.from([1, 2, 3]));

  let type: DataType;
  if (field.type === FieldType.number) {
    type = new Float64();
  } else if (field.type === FieldType.time) {
    type = new TimestampMillisecond();
  } else if (field.type === FieldType.boolean) {
    type = new Bool();
  } else if (field.type === FieldType.string) {
    type = new Utf8();
  } else {
    type = new Utf8();
  }
  const builder = Builder.new({ type, nullValues: [null] });
  field.values.toArray().forEach(builder.append.bind(builder));
  return builder.finish().toVector();
}

export function grafanaDataFrameToArrowTable(data: DataFrame): Table {
  const table = Table.new(
    data.fields.map(field => {
      const column = Column.new(field.name, toArrowVector(field));
      if (field.labels) {
        column.metadata.set('labels', JSON.stringify(field.labels));
      }
      if (field.config) {
        column.metadata.set('config', JSON.stringify(field.config));
      }
      return column;
    })
  );
  const metadata = table.schema.metadata;
  if (data.name) {
    metadata.set('name', data.name);
  }
  if (data.refId) {
    metadata.set('refId', data.refId);
  }
  if (data.meta) {
    metadata.set('meta', JSON.stringify(data.meta));
  }
  return table;
}

export function resultsToDataFrames(rsp: any): DataFrame[] {
  const frames: DataFrame[] = [];
  for (const res of Object.values(rsp.results)) {
    for (const b of (res as any).dataframes) {
      const t = base64StringToArrowTable(b as string);
      frames.push(arrowTableToDataFrame(t));
    }
  }
  return frames;
}
