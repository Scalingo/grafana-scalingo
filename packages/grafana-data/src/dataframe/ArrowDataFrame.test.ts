import fs from 'fs';
import path from 'path';

import { resultsToDataFrames, grafanaDataFrameToArrowTable, arrowTableToDataFrame } from './ArrowDataFrame';
import { toDataFrameDTO, toDataFrame } from './processDataFrame';
import { FieldType } from '../types';
import { Table } from 'apache-arrow';

/* eslint-disable */
const resp = {
  results: {
    '': {
      refId: '',
      dataframes: [
        'QVJST1cxAACsAQAAEAAAAAAACgAOAAwACwAEAAoAAAAUAAAAAAAAAQMACgAMAAAACAAEAAoAAAAIAAAAUAAAAAIAAAAoAAAABAAAAOD+//8IAAAADAAAAAIAAABHQwAABQAAAHJlZklkAAAAAP///wgAAAAMAAAAAAAAAAAAAAAEAAAAbmFtZQAAAAACAAAAlAAAAAQAAACG////FAAAAGAAAABgAAAAAAADAWAAAAACAAAALAAAAAQAAABQ////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAAB0////CAAAAAwAAAAAAAAAAAAAAAQAAABuYW1lAAAAAAAAAABm////AAACAAAAAAAAABIAGAAUABMAEgAMAAAACAAEABIAAAAUAAAAbAAAAHQAAAAAAAoBdAAAAAIAAAA0AAAABAAAANz///8IAAAAEAAAAAQAAAB0aW1lAAAAAAQAAAB0eXBlAAAAAAgADAAIAAQACAAAAAgAAAAQAAAABAAAAFRpbWUAAAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAEAAAAVGltZQAAAAC8AAAAFAAAAAAAAAAMABYAFAATAAwABAAMAAAA0AAAAAAAAAAUAAAAAAAAAwMACgAYAAwACAAEAAoAAAAUAAAAWAAAAA0AAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABoAAAAAAAAAGgAAAAAAAAAAAAAAAAAAABoAAAAAAAAAGgAAAAAAAAAAAAAAAIAAAANAAAAAAAAAAAAAAAAAAAADQAAAAAAAAAAAAAAAAAAAAAAAAAAFp00e2XHFQAIo158ZccVAPqoiH1lxxUA7K6yfmXHFQDetNx/ZccVANC6BoFlxxUAwsAwgmXHFQC0xlqDZccVAKbMhIRlxxUAmNKuhWXHFQCK2NiGZccVAHzeAohlxxUAbuQsiWXHFQAAAAAAAAhAAAAAAAAACEAAAAAAAAAIQAAAAAAAABRAAAAAAAAAFEAAAAAAAAAUQAAAAAAAAAhAAAAAAAAACEAAAAAAAAAIQAAAAAAAABRAAAAAAAAAFEAAAAAAAAAUQAAAAAAAAAhAEAAAAAwAFAASAAwACAAEAAwAAAAQAAAALAAAADgAAAAAAAMAAQAAALgBAAAAAAAAwAAAAAAAAADQAAAAAAAAAAAAAAAAAAAAAAAKAAwAAAAIAAQACgAAAAgAAABQAAAAAgAAACgAAAAEAAAA4P7//wgAAAAMAAAAAgAAAEdDAAAFAAAAcmVmSWQAAAAA////CAAAAAwAAAAAAAAAAAAAAAQAAABuYW1lAAAAAAIAAACUAAAABAAAAIb///8UAAAAYAAAAGAAAAAAAAMBYAAAAAIAAAAsAAAABAAAAFD///8IAAAAEAAAAAYAAABudW1iZXIAAAQAAAB0eXBlAAAAAHT///8IAAAADAAAAAAAAAAAAAAABAAAAG5hbWUAAAAAAAAAAGb///8AAAIAAAAAAAAAEgAYABQAEwASAAwAAAAIAAQAEgAAABQAAABsAAAAdAAAAAAACgF0AAAAAgAAADQAAAAEAAAA3P///wgAAAAQAAAABAAAAHRpbWUAAAAABAAAAHR5cGUAAAAACAAMAAgABAAIAAAACAAAABAAAAAEAAAAVGltZQAAAAAEAAAAbmFtZQAAAAAAAAAAAAAGAAgABgAGAAAAAAADAAQAAABUaW1lAAAAANgBAABBUlJPVzE=',
        'QVJST1cxAAC8AQAAEAAAAAAACgAOAAwACwAEAAoAAAAUAAAAAAAAAQMACgAMAAAACAAEAAoAAAAIAAAAUAAAAAIAAAAoAAAABAAAAND+//8IAAAADAAAAAIAAABHQgAABQAAAHJlZklkAAAA8P7//wgAAAAMAAAAAAAAAAAAAAAEAAAAbmFtZQAAAAACAAAApAAAAAQAAAB2////FAAAAGgAAABoAAAAAAADAWgAAAACAAAALAAAAAQAAABA////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAABk////CAAAABQAAAAJAAAAR0Itc2VyaWVzAAAABAAAAG5hbWUAAAAAAAAAAF7///8AAAIACQAAAEdCLXNlcmllcwASABgAFAATABIADAAAAAgABAASAAAAFAAAAGwAAAB0AAAAAAAKAXQAAAACAAAANAAAAAQAAADc////CAAAABAAAAAEAAAAdGltZQAAAAAEAAAAdHlwZQAAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAQAAABUaW1lAAAAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMABAAAAFRpbWUAAAAAvAAAABQAAAAAAAAADAAWABQAEwAMAAQADAAAANAAAAAAAAAAFAAAAAAAAAMDAAoAGAAMAAgABAAKAAAAFAAAAFgAAAANAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaAAAAAAAAABoAAAAAAAAAAAAAAAAAAAAaAAAAAAAAABoAAAAAAAAAAAAAAACAAAADQAAAAAAAAAAAAAAAAAAAA0AAAAAAAAAAAAAAAAAAAAAAAAAABadNHtlxxUACKNefGXHFQD6qIh9ZccVAOyusn5lxxUA3rTcf2XHFQDQugaBZccVAMLAMIJlxxUAtMZag2XHFQCmzISEZccVAJjSroVlxxUAitjYhmXHFQB83gKIZccVAG7kLIllxxUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAABAAAAAMABQAEgAMAAgABAAMAAAAEAAAACwAAAA4AAAAAAADAAEAAADIAQAAAAAAAMAAAAAAAAAA0AAAAAAAAAAAAAAAAAAAAAAACgAMAAAACAAEAAoAAAAIAAAAUAAAAAIAAAAoAAAABAAAAND+//8IAAAADAAAAAIAAABHQgAABQAAAHJlZklkAAAA8P7//wgAAAAMAAAAAAAAAAAAAAAEAAAAbmFtZQAAAAACAAAApAAAAAQAAAB2////FAAAAGgAAABoAAAAAAADAWgAAAACAAAALAAAAAQAAABA////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAABk////CAAAABQAAAAJAAAAR0Itc2VyaWVzAAAABAAAAG5hbWUAAAAAAAAAAF7///8AAAIACQAAAEdCLXNlcmllcwASABgAFAATABIADAAAAAgABAASAAAAFAAAAGwAAAB0AAAAAAAKAXQAAAACAAAANAAAAAQAAADc////CAAAABAAAAAEAAAAdGltZQAAAAAEAAAAdHlwZQAAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAQAAABUaW1lAAAAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMABAAAAFRpbWUAAAAA6AEAAEFSUk9XMQ==',
      ],
      series: [] as any[],
      tables: null as any,
      frames: null as any,
    },
  },
};
/* eslint-enable */

describe('GEL Utils', () => {
  test('should parse output with dataframe', () => {
    const frames = resultsToDataFrames(resp);
    for (const frame of frames) {
      console.log('Frame', frame.refId);
      for (const field of frame.fields) {
        console.log(' > ', field.name, field.labels);
        console.log(' (values)= ', field.values.toArray());
      }
    }

    const norm = frames.map(f => toDataFrameDTO(f));
    expect(norm).toMatchSnapshot();
  });

  test('processEmptyResults', () => {
    const frames = resultsToDataFrames({
      results: { '': { refId: '', meta: null, series: null, tables: null, dataframes: null } },
    });
    expect(frames.length).toEqual(0);
  });
});

describe('Read/Write arrow Table to DataFrame', () => {
  test('should parse output with dataframe', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      meta: {
        aaa: 'xyz',
        anything: 'xxx',
      },
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'value', config: { min: 0, max: 50, unit: 'somthing' }, type: FieldType.number, values: [1, 2, 3] },
        { name: 'str', config: {}, type: FieldType.string, values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should support duplicate field names', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'a', values: [1, 2, 3] },
        { name: 'a', values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should read all types', () => {
    const fullpath = path.resolve(__dirname, './__snapshots__/all_types.golden.arrow');
    const arrow = fs.readFileSync(fullpath);
    const table = Table.from([arrow]);
    const frame = arrowTableToDataFrame(table);
    expect(toDataFrameDTO(frame)).toMatchSnapshot();
  });
});
