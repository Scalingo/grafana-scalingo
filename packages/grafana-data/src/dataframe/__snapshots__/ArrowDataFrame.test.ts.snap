// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GEL Utils should parse output with dataframe 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "Time",
        "type": "time",
        "values": Array [
          1569334575000,
          1569334580000,
          1569334585000,
          1569334590000,
          1569334595000,
          1569334600000,
          1569334605000,
          1569334610000,
          1569334615000,
          1569334620000,
          1569334625000,
          1569334630000,
          1569334635000,
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "",
        "type": "number",
        "values": Array [
          3,
          3,
          3,
          5,
          5,
          5,
          3,
          3,
          3,
          5,
          5,
          5,
          3,
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": "GC",
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "Time",
        "type": "time",
        "values": Array [
          1569334575000,
          1569334580000,
          1569334585000,
          1569334590000,
          1569334595000,
          1569334600000,
          1569334605000,
          1569334610000,
          1569334615000,
          1569334620000,
          1569334625000,
          1569334630000,
          1569334635000,
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "GB-series",
        "type": "number",
        "values": Array [
          0,
          0,
          0,
          2,
          2,
          2,
          0,
          0,
          0,
          2,
          2,
          2,
          0,
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": "GB",
  },
]
`;

exports[`Read/Write arrow Table to DataFrame should read all types 1`] = `
Object {
  "fields": Array [
    Object {
      "config": Object {},
      "labels": Object {
        "aLabelKey": "aLabelValue",
      },
      "name": "string_values",
      "type": "string",
      "values": Array [
        "Go Min",
        "JS Min (for >= 64)",
        "0 / nil / misc",
        "JS Max (for >= 64)",
        "Go Max",
      ],
    },
    Object {
      "config": Object {
        "decimals": 2,
        "filterable": false,
        "links": Array [
          Object {
            "targetBlank": true,
            "title": "Donate - The Sloth Conservation Foundation",
            "url": "https://slothconservation.com/how-to-help/donate/",
          },
        ],
        "max": null,
        "min": null,
        "noValue": "üò§",
        "nullValueMode": "null",
        "title": "Grafana ‚ù§Ô∏è (Previous should be heart emoji) ü¶• (Previous should be sloth emoji)",
      },
      "labels": Object {
        "aLabelKey": "aLabelValue",
        "bLabelKey": "bLabelValue",
      },
      "name": "nullable_string_values",
      "type": "string",
      "values": Array [
        "Grafana",
        "‚ù§Ô∏è",
        null,
        "ü¶•",
        "update your unicode/font if no sloth, is 2019.",
      ],
    },
    Object {
      "config": Object {
        "max": 1,
        "min": 0,
      },
      "labels": undefined,
      "name": "int8_values",
      "type": "number",
      "values": Array [
        -128,
        -128,
        0,
        127,
        127,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int8_values",
      "type": "number",
      "values": Array [
        -128,
        -128,
        null,
        127,
        127,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int16_values",
      "type": "number",
      "values": Array [
        -32768,
        -32768,
        0,
        32767,
        32767,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int16_values",
      "type": "number",
      "values": Array [
        -32768,
        -32768,
        null,
        32767,
        32767,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int32_values",
      "type": "number",
      "values": Array [
        -2147483648,
        -2147483648,
        1,
        2147483647,
        2147483647,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int32_values",
      "type": "number",
      "values": Array [
        -2147483648,
        -2147483648,
        null,
        2147483647,
        2147483647,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int64_values",
      "type": "number",
      "values": Array [
        "\\"-9223372036854775808\\"",
        "\\"-9007199254740991\\"",
        "\\"1\\"",
        "\\"9007199254740991\\"",
        "\\"9223372036854775807\\"",
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int64_values",
      "type": "number",
      "values": Array [
        "\\"-9223372036854775808\\"",
        "\\"-9007199254740991\\"",
        null,
        "\\"9007199254740991\\"",
        "\\"9223372036854775807\\"",
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint8_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        255,
        255,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint8_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        255,
        255,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint16_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        65535,
        65535,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint16_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        65535,
        65535,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint32_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        4294967295,
        4294967295,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint32_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        4294967295,
        4294967295,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint64_values",
      "type": "number",
      "values": Array [
        "\\"0\\"",
        "\\"0\\"",
        "\\"1\\"",
        "\\"9007199254740991\\"",
        "\\"18446744073709551615\\"",
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint64_values",
      "type": "number",
      "values": Array [
        "\\"0\\"",
        "\\"0\\"",
        null,
        "\\"9007199254740991\\"",
        "\\"18446744073709551615\\"",
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "float32_values",
      "type": "number",
      "values": Array [
        1.401298464324817e-45,
        1.401298464324817e-45,
        1,
        3.4028234663852886e+38,
        3.4028234663852886e+38,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_float32_values",
      "type": "number",
      "values": Array [
        1.401298464324817e-45,
        1.401298464324817e-45,
        null,
        3.4028234663852886e+38,
        3.4028234663852886e+38,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "float64_values",
      "type": "number",
      "values": Array [
        5e-324,
        -9007199254740991,
        1,
        9007199254740991,
        1.7976931348623157e+308,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_float64_values",
      "type": "number",
      "values": Array [
        5e-324,
        -9007199254740991,
        null,
        1.7976931348623157e+308,
        9007199254740991,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "bool_values",
      "type": "boolean",
      "values": Array [
        true,
        false,
        true,
        true,
        false,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_bool_values",
      "type": "boolean",
      "values": Array [
        true,
        false,
        null,
        true,
        false,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "timestamps",
      "type": "time",
      "values": Array [
        0,
        1568039445000,
        1568039450000,
        9007199254.740992,
        9223372036854.775,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_timestamps",
      "type": "time",
      "values": Array [
        0,
        1568039445000,
        null,
        9007199254.740992,
        9223372036854.775,
      ],
    },
  ],
  "meta": Object {
    "limit": 4242,
    "searchWords": Array [
      "Grafana",
      "‚ù§Ô∏è",
      " ü¶• ",
      "test",
    ],
  },
  "name": "many_types",
  "refId": "A",
}
`;
